pennylane_lightning/src/util/Memory.hpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdlib>
+#include <limits>
+#include <memory>
+#include <new>
+
+#include "BitUtil.hpp"
+#include "TypeList.hpp"
+
+namespace Pennylane::Util {
+/**
+ * @brief Custom aligned allocate function. As appleclang does not support
+ * std::aligned_alloc in Mac OS 10.14, we use posix_memalign function.
+ *
+ * Note that alignment must be larger than max_align_t. Otherwise, the behavior
+ * is undefined.
+ *
+ * @param alignment Alignment value we want for the data pointer
+ * @param bytes Number of bytes to allocate
+ * @return Memory pointer
+ */
+inline auto alignedAlloc(uint32_t alignment, size_t bytes) -> void * {
+#if defined(__clang__)
+    /* Apple clang does not support std::aligned_alloc in Mac 10.14.
+     * Thus we use Posix function instead. */
+    void *p;
+    posix_memalign(&p, alignment, bytes);
+    return p;
+#elif defined(_MSC_VER)
+    return _aligned_malloc(bytes, alignment);
```

Comment:
Interesting too that Windows does not default to cpp standard here.

Reply:
Yes, the underlying reason is that the standard requires that even aligned memory can be deallocated with `std:free`, but this is impossible on Windows. The alignment business is quite tricky yet.